Interactivity: Fix string index warning

sirreal opened the pull request:
Fix the following warning when using an interactivity bind directive with a short attribute name, e.g. `wp-data-bind--id`.

> Warning: Uninitialized string offset 4 in /var/www/html/wp-includes/interactivity-api/class-wp-interactivity-api.php on line 579


Trac ticket: https://core.trac.wordpress.org/ticket/60758

---
**This Pull Request is for code review only. Please keep all other discussion in the Trac ticket. Do not merge this Pull Request. See [GitHub Pull Requests for Code Review](https://make.wordpress.org/core/handbook/contribute/git/github-pull-requests-for-code-review/) in the Core Handbook for more details.**

sirreal committed:
Fix out-of-bounds string index warning

sirreal commented:
👋 @c4rl0sbr4v0 @DAreRodz for review.

sirreal commented on src/wp-includes/interactivity-api/class-wp-interactivity-api.php:
```diff
@@ -576,7 +576,10 @@ private function data_wp_bind_processor( WP_Interactivity_API_Directives_Process
 
-				if ( null !== $result && ( false !== $result || '-' === $bound_attribute[4] ) ) {
+				if ( null !== $result && (
+						false !== $result ||
+						( strlen( $bound_attribute ) > 5 && '-' === $bound_attribute[4] )
```
Bad indentation
```suggestion
					false !== $result ||
					( strlen( $bound_attribute ) > 5 && '-' === $bound_attribute[4] )
```

sirreal committed:
Fix indentation

cbravobernal reviewed (approved):
LGTM

swissspidy commented:
Possible to add tests for this?

sirreal commented:
I tried adding tests and would like to, but I couldn't get them to fail on the warning, that may be a configuration option or my lack of knowledge, but I couldn't find any way of saying "warnings on _this test_ should cause the test to fail."

There is actually a test that has a short attribute and should hit this code path, but it doesn't fail the test suite or seem to generate any warning:

https://github.com/WordPress/wordpress-develop/blob/a6783cfbee4f72b38d67744c2a456b89dce626ab/tests/phpunit/tests/interactivity-api/wpInteractivityAPI-wp-bind.php#L368-L381

gziolo commented:
I can see the error when running tests. I might be valuable to add another test that handles the boolean value:

<img width="1243" alt="Screenshot 2024-03-14 at 12 39 34" src="https://github.com/WordPress/wordpress-develop/assets/699132/c12bd5a0-cf88-480a-8795-d64d1a061c7a">


The same check needs to be added in another line which only will trigger when passing `true`:

```php
if ( is_bool( $result ) && strlen( $bound_attribute ) > 5 && '-' === $bound_attribute[4] ) {
```

sirreal committed:
Fix another instance of the same warning

sirreal commented:
Good catch! I pushed a change to apply the same fix to that warning.

gziolo commented:
You also need a test like the one I shared to trigger the warning in the second updated line. The value needs to be `true`.

sirreal committed:
Add true value test

sirreal committed:
Fix test name

sirreal commented:
> You also need a test like the one I shared to trigger the warning in the second updated line. The value needs to be `true`.

I've pushed a test for this, but I don't see the warnings when running the tests 😕

sirreal commented:
I pushed a test that does error if I revert the changes.

sirreal commented:
I'll address the test failures.

sirreal committed:
Tests should be using assertSame

sirreal committed:
Attribute is set to true value

sirreal commented:
I fixed the tests and updated them to use `assertSame`. We're comparing values of types we expect to match, mostly strings, so `assertSame` is the appropriate assertion.

This is something to address across the interactivity tests when comparing strings to strings.

gziolo reviewed (approved):
Excellent, tests pass locally. I no longer see the warnings. I could confirm that unit tests hit both conditions by manually changing the comparisons `strlen( $bound_attribute ) > 5` using value `1`.

swissspidy commented:
Committed in https://core.trac.wordpress.org/changeset/57835
